org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/discovery/ZigBeeDiscoveryParticipant.java

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.zigbee.discovery;
+
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+
+import com.zsmartsystems.zigbee.ZigBeeNode;
+
+/**
+ * A {@link ZigBeeDiscoveryParticipant} that is registered as a service is picked up by the ZigBeeDiscoveryService
+ * and can contribute {@link DiscoveryResult}s from ZigBee scans.
+ *
+ * @author Chris Jackson
+ *
+ */
+@NonNullByDefault
+public interface ZigBeeDiscoveryParticipant {
+
+    /**
+     * Defines the list of thing types that this participant can identify
+     *
+     * @return a set of thing type UIDs for which discovery results can be created
+     */
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs();
+
+    /**
+     * Returns the thing type UID for a ZigBee device
+     *
+     * @param node the {@link ZigBeeNode} found on the network
+     * @return a thing type UID or <code>null</code>, if the device is not supported by this participant
+     */
+    @Nullable
+    public ThingTypeUID getThingTypeUID(ZigBeeNode node);
+
+    /**
+     * Creates a discovery result for a ZigBee device. Note that the {@link DiscoveryResult} must contain the properties
+     * provided, but can contain additional properties as needed by the thing.
+     *
+     * @param bridgeUID the {@link ThingUID} of the bridge on which this node was discovered
+     * @param node the {@link ZigBeeNode} found on the network
+     * @param properties a map of properties that must be added to the {@link DiscoveryResult}
+     * @return the discovery result or <code>null</code>, if the node is not supported by this participant
+     */
+    @Nullable
+    public DiscoveryResult createResult(ThingUID bridgeUID, ZigBeeNode node, Map<String, Object> properties);
```

Comment:
When seeing this, at first I found it somewhat peculiar that the `ZigBeeDiscoveryParticipant` is responsible for adding the `properties` to the `DiscoveryResult` - and then realized that one cannot add properties to a `DiscoveryResult` after it has been created.

Just as an idea, one could also change the interface here and require participants to return a `DiscoveryResultBuilder`, to which the properties could then be added by the `ZigBeeDiscoveryService`. Advantage of this solution would be that the binding can make sure that all required properties are set.

I think it would be a small improvement - but is ok as is as well.

Reply:
I don't think that works as it does not then allow the participant to add properties in addition to the properties that the discovery service wants to add - ie if you return the ```DiscoveryResultBuilder``` then you still can't access the properties that were added already - you can only replace them.

Reply:
I have looked at the `DiscoveryResultBuilder` - one cannot replace properties, but one can augment the set of properties.

The JavaDoc of `DiscoveryResultBuilder#withProperties` says: " Adds properties to the desired result", and this is also reflected in its implementation, where `this.properties.putAll(properties);` is used.

I.e., using a `DiscoveryResultBuilder` here should work.

Reply:
Ah ok, I missed this point. I don't mind - neither way is especially nice IMHO. I had originally hoped that I could update the DiscoveryResult with additional properties, but this is not possible.  I'll have a think - maybe your suggestion sounds nicer...